cmd/counters/src/lib.rs

```diff
@@ -0,0 +1,281 @@
+                // counters struct; this check is imprecise but probably good enough
+                if s.name.contains("Counts")
+                    || s.name.contains("CountedRingbuf")
+                {
+                    Some(s)
+                } else {
+                    continue;
+                }
+            }
+            Err(_) => {
+                // Type lookup failed, so fall back to the variable name
+                if varname.ends_with("COUNTERS") {
+                    None
+                } else {
+                    continue;
+                }
+            }
+        };
+        let t = taskname(hubris, var)?;
+        if let Some(ref name) = subargs.name {
+            if varname.contains(name) || t.contains(name) {
+                counters.entry(t).or_default().push(((varname, var), def));
+            }
+        } else {
+            counters.entry(t).or_default().push(((varname, var), def));
+        }
+    }
+
+    if counters.is_empty() {
+        if let Some(name) = subargs.name {
+            bail!(
+                "no counters found with names containing \"{}\" (-l to list)",
+                name
+            );
+        } else {
+            bail!("no counters found");
+        }
+    }
+
+    for vars in counters.values_mut() {
+        vars.sort_by_key(|&(v, _)| v);
+    }
+
+    if subargs.list {
+        for (t, ctrs) in counters {
+            println!("{t}:");
+            println!("    {:<10} {:>5} VARIABLE", "ADDR", "SIZE",);
+            for ((name, var), _) in ctrs {
+                println!("    {:<#010x} {:>5} {name}", var.addr, var.size)
+            }
+        }
+
+        return Ok(());
+    }
+
+    for (t, ctrs) in counters {
+        // Try not to use `?` here, because it causes one bad counter to make
+        // them all unavailable.
+        println!("{t}\n |");
+        let mut ctrs = ctrs.iter().peekable();
+        while let Some(((varname, var), def)) = ctrs.next() {
+            if let Some(def) = def {
+                println!(" +---> {varname}:");
+                let pad = if ctrs.peek().is_some() { " |  " } else { "    " };
+                if let Err(e) =
+                    counter_dump(hubris, core, def, var, &subargs, pad)
+                {
+                    if subargs.verbose {
+                        humility::msg!("counter dump failed: {e:?}");
+                    } else {
+                        humility::msg!("counter dump failed: {e}");
+                    }
+                }
+            } else {
+                humility::msg!("could not look up type: {:?}", var.goff);
+            }
+        }
+    }
+    Ok(())
+}
+
+fn counter_dump(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    def: &HubrisStruct,
+    var: &HubrisVariable,
+    args: &CountersArgs,
+    pad: &str,
+) -> Result<()> {
+    let mut buf: Vec<u8> = vec![];
+    buf.resize_with(var.size, Default::default);
```

Comment:
Out of curiosity, is this better than
```rust
let mut buf: Vec<u8> = vec![Default::default(); var.size]
```
or even
```rust
let mut buf = vec![0u8; var.size];
```
?

Reply:
i copied this from `humility ringbuf`, so the only reason i can give you is "for historical reasons". i think your code is nicer, so i've changed it to do that.